contracts/cw-ics721-bridge/src/msg.rs

```diff
@@ -0,0 +1,189 @@
+    },
+    /// Much like mint, but will instantiate a new cw721 contract iff
+    /// the classID does not have one yet.
+    DoInstantiateAndMint {
+        /// The ics721 class ID to mint for.
+        class_id: String,
+        /// The URI for this class ID.
+        class_uri: Option<String>,
+        /// Unique identifiers for the tokens being transfered.
+        token_ids: Vec<String>,
+        /// A list of urls pointing to metadata about the NFTs. For
+        /// example, this may point to ERC721 metadata on ipfs.
+        ///
+        /// Must be the same length as token_ids.
+        token_uris: Vec<String>,
+        /// The address that ought to receive the NFT. This is a local
+        /// address, not a bech32 public key.
+        receiver: String,
+    },
+    /// Transfers a number of NFTs identified by CLASS_ID and
+    /// TOKEN_IDS to RECEIVER.
+    BatchTransfer {
+        /// The ics721 class ID of the tokens to be transfered.
+        class_id: String,
+        /// The address that should receive the tokens.
+        receiver: String,
+        /// The tokens (of CLASS_ID) that should be sent.
+        token_ids: Vec<String>,
+    },
+    /// Handles the falliable part of receiving an IBC
+    /// packet. Transforms TRANSFERS into a `BatchTransfer` message
+    /// and NEW_TOKENS into a `DoInstantiateAndMint`, then dispatches
+    /// those methods.
+    HandlePacketReceive {
+        /// The address receiving the NFTs.
+        receiver: String,
+        /// The URI for the collection being transfered.
+        class_uri: Option<String>,
+        /// Information about transfer actions.
+        transfers: Option<TransferInfo>,
+        /// Information about mint actions.
+        new_tokens: Option<NewTokenInfo>,
+    },
+}
+
+#[cw_serde]
+pub struct TransferInfo {
+    /// The class ID the tokens belong to.
+    pub class_id: String,
+    /// The tokens to be transfered.
+    pub token_ids: Vec<String>,
+}
+
+#[cw_serde]
+pub struct NewTokenInfo {
+    /// The class ID to mint tokens for.
+    pub class_id: String,
+    /// The token IDs of the tokens to be minted.
+    pub token_ids: Vec<String>,
+    /// The URIs of the tokens to be minted. Matched with token_ids by
+    /// index.
+    pub token_uris: Vec<String>,
+}
+
+#[cw_serde]
+pub struct IbcAwayMsg {
+    /// The address that should receive the NFT being sent on the
+    /// *receiving chain*.
+    pub receiver: String,
+    /// The *local* channel ID this ought to be sent away on. This
+    /// contract must have a connection on this channel.
+    pub channel_id: String,
+    /// Timeout for the IBC message. TODO: make this optional and set
+    /// default?
+    pub timeout: IbcTimeout,
+}
+
+#[cw_serde]
+#[derive(QueryResponses)]
+pub enum QueryMsg {
+    /// Gets the classID this contract has stored for a given NFT
+    /// contract. If there is no class ID for the provided contract,
+    /// returns None.
+    #[returns(Option<String>)]
```

Comment:
I'd rather encapsulate return types in a response struct like `ClassIdResponse`. The `...ForX` in these is a redundant since the `X` is already referenced in the field. Response types are more extensible as they can grow as more fields are added. They're also easier to handle on the frontend and follows CosmWasm conventions thus far.

Reply:
good point about the query names, i've updated that.

i will push back against the response types a little though. at least in my experience writing the dao dao frontend, returning "regular" types is actually more ergonomic than having to index into `response.thing_you_care_about` every time. additionally, we can get the same typed abstractions from ts-codegen ([dao-contracts example](https://github.com/DA0-DA0/dao-contracts/blob/main/typescript/contracts/cwd-proposal-single/CwdProposalSingleContract.ts#L375)).

i also find wrapping things in response types to be markedly less readable on the contract side. if something returns `Addr`, i know what that means. if something returns `AdminResponse` i find the layer of indirection strange as i expect it to mean something.

ultimately, your call. strong opinion weakly held.

Reply:
Okay lets go with this and see how it works out. I tend to optimize for least developer frustration (LDF -- I made that up). I can imagine a frontend developer would spend a couple of cycles figuring out that some queries return results in an encapsulated response type and others don't (i.e: `OwnerOfResponse` vs `Option<Addr>`).